framework/model/metric/accesslog_convertor.go

```diff
@@ -0,0 +1,103 @@
+package metric
+
+import (
+	"errors"
+	data_accesslog "github.com/envoyproxy/go-control-plane/envoy/data/accesslog/v3"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+	"sync"
+)
+
+type AccessLogConvertor struct {
+	name            string                       // handler name
+	cacheResult     map[string]map[string]string // meta -> value
+	cacheResultCopy map[string]map[string]string
+	handler         func(logEntry []*data_accesslog.HTTPAccessLogEntry, clientSet *kubernetes.Clientset) (map[string]map[string]string, error)
+	convertorLock   sync.RWMutex
+	clientSet       *kubernetes.Clientset
+}
+
+func NewAccessLogConvertor(config AccessLogConvertorConfig) *AccessLogConvertor {
+	return &AccessLogConvertor{
+		name:            config.Name,
+		clientSet:       config.ClientSet,
+		handler:         config.Handler,
+		cacheResult:     make(map[string]map[string]string),
+		cacheResultCopy: make(map[string]map[string]string),
+	}
+}
+
+func (alc *AccessLogConvertor) Name() string {
+	return alc.name
+}
+
+func (alc *AccessLogConvertor) CacheResultCopy() map[string]map[string]string {
+	alc.convertorLock.RLock()
+	defer alc.convertorLock.RUnlock()
+
+	return alc.cacheResultCopy
+}
+
+func (alc *AccessLogConvertor) Convert(logEntry []*data_accesslog.HTTPAccessLogEntry) error {
+	log := log.WithField("reporter", "AccessLogConvertor").WithField("function", "Convert")
+	tmpResult, err := alc.handler(logEntry, alc.clientSet)
+	if err != nil {
+		return err
+	}
+
+	alc.convertorLock.Lock()
+	defer alc.convertorLock.Unlock()
+
+	if len(tmpResult) == 0 {
+		return errors.New("tmpResult is nil")
+	}
+
+	needUpdate := false
+	// merge result, copy on write
+	for meta, value := range tmpResult {
+
+		if _, ok := alc.cacheResult[meta]; !ok {
+			// new meta
+			log.Debugf("alc.cacheResult[%s] is not ok, inits", meta)
+			needUpdate = true
+			alc.cacheResult[meta] = make(map[string]string)
+			for k, v := range value {
+				alc.cacheResult[meta][k] = v
+			}
+			continue
+		}
+
+		// existed meta
+		if !needUpdate {
+			needUpdate = valueMerge(alc.cacheResult[meta], value)
+		} else {
+			valueMerge(alc.cacheResult[meta], value)
+		}
+
+	}
+
+	if needUpdate {
+		newCacheResultCopy := make(map[string]map[string]string)
+		for meta, value := range alc.cacheResult {
+			newCacheResultCopy[meta] = make(map[string]string)
+			for k, v := range value {
+				newCacheResultCopy[meta][k] = v
+			}
+		}
+		alc.cacheResultCopy = newCacheResultCopy
+	}
+
+	return err
+}
+
+func valueMerge(cacheValue, tmpValue map[string]string) bool {
+	needMerge := false
+	for k, v := range tmpValue {
+		// new key
+		if _, ok := cacheValue[k]; !ok {
```

Comment:
new key or update?

Reply:
Key is destination service, value is the calls number. So value is useless when key is existed.